src/celengine/astrodb.cpp

```diff
@@ -0,0 +1,157 @@
+{
+    MainIndex::const_iterator it = m_mainIndex.find(nr);
+    if (it == m_mainIndex.end())
+        return nullptr;
+    return it->second;
+}
+
+AstroCatalog::IndexNumber AstroDatabase::searchCrossIndexForCatalogNumber(int catalog, AstroCatalog::IndexNumber nr) const
+{
+    std::unordered_map<int, CrossIndex*>::const_iterator it = m_catxindex.find(catalog);
+    if (it == m_catxindex.end())
+        return AstroCatalog::InvalidIndex;
+    if (it->second->count(nr) > 0)
+        return it->second->at(nr);
+    return AstroCatalog::InvalidIndex;
+}
+
+AstroCatalog::IndexNumber AstroDatabase::crossIndex(int catalog, AstroCatalog::IndexNumber nr) const
+{
+    std::unordered_map<int, CrossIndex*>::const_iterator it = m_celxindex.find(catalog);
+    if (it == m_celxindex.end())
+        return AstroCatalog::InvalidIndex;
+    if (it->second->count(nr) > 0)
+        return it->second->at(nr);
+    return AstroCatalog::InvalidIndex;
+}
+
+bool AstroDatabase::isInCrossIndex(int catalog, AstroCatalog::IndexNumber nr) const
+{
+    std::unordered_map<int, CrossIndex*>::const_iterator it = m_catxindex.find(catalog);
+    if (it == m_catxindex.end())
+        return false;
+    if (it->second->count(nr) == 0)
+        return false;
+    return true;
+}
+
+AstroCatalog::IndexNumber AstroDatabase::findCatalogNumberByName(const std::string& name) const
+{
+    AstroCatalog::IndexNumber nr = m_nameDB.getCatalogNumberByName(name);
+    if (nr != AstroCatalog::InvalidIndex)
+        return nr;
+    for (const auto ci : m_catalogs)
+    {
+        AstroCatalog::IndexNumber inr = ci.second->getIndexNumberByName(name);
+        if (inr == AstroCatalog::InvalidIndex)
+            continue;
+        if (ci.first == Celestia)
+            nr = inr;
+        else
+            nr = searchCrossIndexForCatalogNumber(ci.first, inr);
+        if (nr != AstroCatalog::InvalidIndex)
+            return nr;
+    }
+    return AstroCatalog::InvalidIndex;
+}
+
+std::string AstroDatabase::catalogNumberToString(AstroCatalog::IndexNumber nr) const
+{
+    return m_catalogs.find(Celestia)->second->getNameByIndexNumber(nr);
+}
+
+std::string AstroDatabase::catalogNumberToString(int catalog, AstroCatalog::IndexNumber nr) const
+{
+    if (m_catalogs.count(catalog) == 0)
+        return "#{Invalid catalog}";
+    return m_catalogs.find(catalog)->second->getNameByIndexNumber(nr);
+}
+
+std::string AstroDatabase::getObjectName(AstroCatalog::IndexNumber nr, bool i18n) const
+{
+    NameDatabase::NumberIndex::const_iterator iter = m_nameDB.getFirstNameIter(nr);
+    if (iter != m_nameDB.getFinalNameIter() && iter->first == nr)
+    {
+        if (i18n && iter->second != _(iter->second.c_str()))
+            return _(iter->second.c_str());
+        else
+            return iter->second;
+    }
+    return catalogNumberToString(nr);
+}
+
+std::string AstroDatabase::getObjectNameList(AstroCatalog::IndexNumber nr, int max) const
+{
+    string names;
```

Comment:
to optimize memory allocations use `string names("", N)` where N is large enough to keep full name, i suppose 127 will be ok. or `string names; names.reserve(N);`

Reply:
With current implementation, such a memory preallocation leads to string with junks at the beginning and normal content after preallocated section.